### PR TITLE
[mlir][python] Make the Context/Operation capsule creation methods work as documented.

### PR DESCRIPTION
**Pull request by stellaraccident**
This fixes a longstanding bug in the `Context._CAPICreate` method whereby it was not taking ownership of the PyMlirContext wrapper when casting to a Python object. The result was minimally that all such contexts transferred in that way would leak. In addition, counter to the documentation for the `_CAPICreate` helper (see `mlir-c/Bindings/Python/Interop.h`) and the `forContext` / `forOperation` methods, we were silently upgrading any unknown context/operation pointer to steal-ownership semantics. This is dangerous and was causing some subtle bugs downstream where this facility is getting the most use.

This patch corrects the semantics and will only do an ownership transfer for `_CAPICreate`, and it will further require that it is an ownership transfer (if already transferred, it was just silently succeeding). Removing the mis-aligned behavior made it clear where the downstream was doing the wrong thing.

It also adds some `_testing_` functions to create unowned context and operation capsules so that this can be fully tested upstream, reworking the tests to verify the behavior.

In some torture testing downstream, I was not able to trigger any memory corruption with the newly enforced semantics. When getting it wrong, a regular exception is raised.

### CODE REVIEW

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 1]
```diff
@@ -615,18 +615,35 @@ PyMlirContextRef PyMlirContext::forContext(MlirContext context) {
...
+  auto it = liveContexts.find(context.ptr);
+  if (it != liveContexts.end()) {
+    throw std::runtime_error(
+        "Cannot transfer ownership of the context to Python "
+        "as it is already owned by Python.");
+  }
+
+  PyMlirContext *unownedContextWrapper = new PyMlirContext(context);
+  // Note that the default return value policy on cast is automatic_reference,
+  // which does not take ownership (delete will not be called).
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRCore.cpp:639]: Just curious: is this specific to `py::cast` and/or a bug in pybind? The docs say about `return_value_policy::automatic`: 

> This policy falls back to the policy return_value_policy::take_ownership when the return value is a pointer... **This is the default policy for `py::class_`-wrapped types**.

and about `return_value_policy::automatic_reference`:

> As above, but use policy `return_value_policy::reference` when the return value is a pointer. This is the default conversion policy for function arguments when calling Python functions manually from C++ code (i.e. via `handle::operator()`) and the casters in `pybind11/stl.h`.

Are we somehow in either of those two cases here?

Comment (stellaraccident) [reply mlir/lib/Bindings/Python/IRCore.cpp:639]: It's somewhat poorly documented -- the behavior is specific to cast whereas the documentation are talking about the more common cases of function args/returns.

Note that I got that comment from the cast in `fromOperation`, which gets it right. That path is used heavily and was fixed long ago. This branch of context creation, though, had no upstream uses and was just wrong.

Looking for the reference on how I know this. I remember researching it some time ago.

**Comment by stellaraccident** [comment]
https://github.com/pybind/pybind11/issues/287#issuecomment-233373721

**Comment by stellaraccident** [comment]
https://github.com/pybind/pybind11/blob/eeac2f45728633d7ee6fe792bea2890345697119/include/pybind11/cast.h#L1110

File: mlir/lib/Bindings/Python/IRCore.cpp (hunk 1 above)
Comment (stellaraccident) [reply mlir/lib/Bindings/Python/IRCore.cpp:639]: Left a couple of links below - explanation by the author and the source.

**Comment by makslevental** [comment]
> https://github.com/pybind/pybind11/blob/eeac2f45728633d7ee6fe792bea2890345697119/include/pybind11/cast.h#L1110

TIL:

> (You can enforce a custom return value policy as a template argument to cast.)

Okay will review with a fine-toothed comb tomorrow!

**Comment by stellaraccident** [comment]
> > [pybind/pybind11#287 (comment)](https://github.com/pybind/pybind11/issues/287#issuecomment-233373721)
> 
> TIL:
> 
> > (You can enforce a custom return value policy as a template argument to cast.)
> 
> Okay will review with a fine-toothed comb tomorrow!

Thanks. I beefed up the downstream tests for this and enabled on all platforms. Afaict, this is the first time this feature has worked reliably: https://github.com/openxla/iree/pull/15982

I think I got the vital characteristics of it tested upstream now, so should be an overall improvement to reliability.

**Review by makslevental — APPROVED** [review]
LGTM - I actually think this will break me in some experimental stuff in mlir-aie but solid fix nonetheless :)

**Comment by stellaraccident** [comment]
> LGTM - I actually think this will break me in some experimental stuff in mlir-aie but solid fix nonetheless :)

If your experience mirrors mine, you may have realized soon that what you were doing was already subtly broken but just hadn't been caught yet :)

**Comment by makslevental** [comment]
> If your experience mirrors mine, you may have realized soon that what you were doing was already subtly broken but just hadn't been caught yet :)

I definitely ran into the occasional ownership issue but I've always been able to brush it under the rug by fiddling with `return_value_policy` 😁. I look forward to trying out this API.

**Comment by ftynse** [comment]
@makslevental  could you give an example of the wrong things downstreams were doing and how to make it right?  This will help downstreams to fix themselves after they bump LLVM version past this commit.

**Comment by makslevental** [comment]
> @makslevental could you give an example of the wrong things downstreams were doing and how to make it right? This will help downstreams to fix themselves after they bump LLVM version past this commit.

Well I'm doing super-gnarly stuff like shipping arbitrary `mlir::Operation`s [across the boundary](https://github.com/Xilinx/mlir-aie/blob/a63eb5dac38c89257c94174d0cb214332083d597/python/PybindTypes.cpp#L37):

```cpp
 return static_cast<PySwitchboxOp &>(
      PyOperation::forOperation(DefaultingPyMlirContext::resolve().getRef(),
                                wrap(switchboxOp.getOperation()))
          ->getOperation());
```

I highly doubt anyone else is doing this (I don't think anyone else knows how to...). `return_value_policy` I've had to reach for when returning arbitrary [C++ functions across the boundary](https://github.com/Xilinx/mlir-aie/blob/a63eb5dac38c89257c94174d0cb214332083d597/python/PybindTypes.cpp#L143) (another thing I doubt anyone else is doing). 

I was actually curious what exactly @stellaraccident was experiencing that prompted this PR (there's a linked PR in iree but it doesn't say much).

> This will help downstreams to fix themselves after they bump LLVM version past this commit.

My plan is to change `forOperation` -> `stealExternalOperation` and see what happens. Not very scientific unfortunately.

**Comment by makslevental** [comment]
Just to be clear: I don't think normal users will be broken by this - I was speaking specifically about myself because I am using `forOperation` outside of `lib/Bindings` (which isn't normally possible).

**Comment by ftynse** [comment]
Yes, without `forOpeation`, users _are_ broken. This basically makes it impossible for "normal" users to return a non-owning reference to an operation. I don't think that's what we want. Here's the minimal example that breaks upstream. Note that anybody downstream cannot return something else than `MlirOperation` AFAIK. Upstream doesn't have a problem because it can, and does, return a `PyOpreationRef`.

```
diff --git a/mlir/lib/Bindings/Python/IRCore.cpp b/mlir/lib/Bindings/Python/IRCore.cpp
index 39757dfad5be..2ce640674245 100644
--- a/mlir/lib/Bindings/Python/IRCore.cpp
+++ b/mlir/lib/Bindings/Python/IRCore.cpp
@@ -3071,6 +3071,11 @@ void mlir::python::populateIRCore(py::module &m) {
                   py::arg("successors") = py::none(), py::arg("regions") = 0,
                   py::arg("loc") = py::none(), py::arg("ip") = py::none(),
                   py::arg("infer_type") = false, kOperationCreateDocstring)
+      .def("_get_first_in_block", [](PyOperation &self) -> MlirOperation {
+        MlirBlock block = mlirOperationGetBlock(self.get());
+        MlirOperation first = mlirBlockGetFirstOperation(block);
+        return first;
+      })
       .def_static(
           "parse",
           [](const std::string &sourceStr, const std::string &sourceName,
diff --git a/mlir/test/python/ir/operation.py b/mlir/test/python/ir/operation.py
index f59b1a26ba48..6b12b8da5c24 100644
--- a/mlir/test/python/ir/operation.py
+++ b/mlir/test/python/ir/operation.py
@@ -24,6 +24,25 @@ def expect_index_error(callback):
     except IndexError:
         pass
 
+@run
+def testCustomBind():
+    ctx = Context()
+    ctx.allow_unregistered_dialects = True
+    module = Module.parse(
+        r"""
+    func.func @f1(%arg0: i32) -> i32 {
+      %1 = "custom.addi"(%arg0, %arg0) : (i32, i32) -> i32
+      return %1 : i32
+    }
+  """,
+        ctx,
+    )
+    add = module.body.operations[0].regions[0].blocks[0].operations[0]
+    op = add.operation
+    # This will get a reference to itself.
+    f1 = op._get_first_in_block()
+
+
 
 # Verify iterator based traversal of the op/region/block hierarchy.
 # CHECK-LABEL: TEST: testTraverseOpRegionBlockIterators
```

**Comment by ftynse** [comment]
Reverting for now. 

I think it's either a matter of (a) updating the caster in PybindAdaptors.h so that it can interpret `MlirOperation` as reference when the corresponding operation itself (and maybe one of its ancestors) is already owned by python or (b) somehow providing a mechanism for downstream binding implementers to indicate whether they intend to transfer ownership. But I'm rusty on this whole ownership business.

**Comment by makslevental** [comment]
> so that it can interpret `MlirOperation` as reference when the corresponding operation itself (and maybe one of its ancestors) is already owned by python or (b) somehow providing a mechanism for downstream binding implementers to indicate whether they intend to transfer ownership.

I see 

```cpp
static handle cast(MlirOperation v, return_value_policy, handle) {
  return py::module::import(MAKE_MLIR_PYTHON_QUALNAME("ir"))
        .attr("Operation")
        .attr(MLIR_PYTHON_CAPI_FACTORY_ATTR)(capsule)
        .release();
}
```

which could be changed to

```cpp
static handle cast(MlirOperation v, return_value_policy p, handle h) {
  // call MLIR_PYTHON_CAPI_FACTORY_ATTR somehow differently depending on p/h
}
```

(`h` is the parent object). 

This would enable downstream users to 

```
.def("_get_first_in_block", [](MlirOperation op) -> MlirOperation {
  MlirBlock block = mlirOperationGetBlock(op);
  MlirOperation first = mlirBlockGetFirstOperation(block);
  return py::cast(first, return_value_policy::reference);
})
```
in order to communicate intent.

But I don't actually know what `// call MLIR_PYTHON_CAPI_FACTORY_ATTR somehow differently depending on p/h` should be - the call to `_CAPICreate` is the important piece and it's a call through the python API (which we don't want to expand to support something like a `return_value_policy` arg?). 

> Upstream doesn't have a problem because it can, and does, return a `PyOpreationRef`.

I'll take this opportunity to beat my favorite dead horse: we could just start exposing the utilities in `lib/Bindings`, such as `PyOpreationRef`, in order to avoid the inherent difficulties of relying on only `PybindAdaptors.h` to always do the right things.

**Comment by stellaraccident** [comment]
> Yes, without `forOpeation`, users _are_ broken. This basically makes it impossible for "normal" users to return a non-owning reference to an operation. I don't think that's what we want. Here's the minimal example that breaks upstream. Note that anybody downstream cannot return something else than `MlirOperation` AFAIK. Upstream doesn't have a problem because it can, and does, return a `PyOpreationRef`.
> 
> ```
> diff --git a/mlir/lib/Bindings/Python/IRCore.cpp b/mlir/lib/Bindings/Python/IRCore.cpp
> index 39757dfad5be..2ce640674245 100644
> --- a/mlir/lib/Bindings/Python/IRCore.cpp
> +++ b/mlir/lib/Bindings/Python/IRCore.cpp
> @@ -3071,6 +3071,11 @@ void mlir::python::populateIRCore(py::module &m) {
>                    py::arg("successors") = py::none(), py::arg("regions") = 0,
>                    py::arg("loc") = py::none(), py::arg("ip") = py::none(),
>                    py::arg("infer_type") = false, kOperationCreateDocstring)
> +      .def("_get_first_in_block", [](PyOperation &self) -> MlirOperation {
> +        MlirBlock block = mlirOperationGetBlock(self.get());
> +        MlirOperation first = mlirBlockGetFirstOperation(block);
> +        return first;
> +      })
>        .def_static(
>            "parse",
>            [](const std::string &sourceStr, const std::string &sourceName,
> diff --git a/mlir/test/python/ir/operation.py b/mlir/test/python/ir/operation.py
> index f59b1a26ba48..6b12b8da5c24 100644
> --- a/mlir/test/python/ir/operation.py
> +++ b/mlir/test/python/ir/operation.py
> @@ -24,6 +24,25 @@ def expect_index_error(callback):
>      except IndexError:
>          pass
>  
> +@run
> +def testCustomBind():
> +    ctx = Context()
> +    ctx.allow_unregistered_dialects = True
> +    module = Module.parse(
> +        r"""
> +    func.func @f1(%arg0: i32) -> i32 {
> +      %1 = "custom.addi"(%arg0, %arg0) : (i32, i32) -> i32
> +      return %1 : i32
> +    }
> +  """,
> +        ctx,
> +    )
> +    add = module.body.operations[0].regions[0].blocks[0].operations[0]
> +    op = add.operation
> +    # This will get a reference to itself.
> +    f1 = op._get_first_in_block()
> +
> +
>  
>  # Verify iterator based traversal of the op/region/block hierarchy.
>  # CHECK-LABEL: TEST: testTraverseOpRegionBlockIterators
> ```

Darn - I was trying to not break that case... Which was the reason for the disjoint implementations between context and op import. Thanks for the test case. That will help in rolling this forward. Probably next week at this point.

**Comment by stellaraccident** [comment]
> Reverting for now. 
> 
> I think it's either a matter of (a) updating the caster in PybindAdaptors.h so that it can interpret `MlirOperation` as reference when the corresponding operation itself (and maybe one of its ancestors) is already owned by python or (b) somehow providing a mechanism for downstream binding implementers to indicate whether they intend to transfer ownership. But I'm rusty on this whole ownership business. 

I might add an arg to indicate whether ownership must be transferred. The case I was running into was subtle non-determinism based on access pattern when a completely external API with its own ownership semantics was negotiating to give ownership to the python API. The implicit ownership capture of context and operation makes this very easy to get wrong with perfectly valid Python sequences. I thought it was a Windows ghost for months but it just turned out to be more sensitive to the fallout. I finally had evidence that a real program was also triggering the issue on Linux (but it was in the bowels of an integration, and what I sent upstream was my reduced test case). That is what made me trade it seriously.

**Comment by stellaraccident** [comment]
> I'll take this opportunity to beat my favorite dead horse: we could just start exposing the utilities in lib/Bindings, such as PyOpreationRef, in order to avoid the inherent difficulties of relying on only PybindAdaptors.h to always do the right things.

Wouldn't help me. I was dealing with something that was using the raw C API via ctypes. The contract here must be built in to the API and enforced correctly.

**Pull request merged**